### PR TITLE
Add DataTableTheme to allow for themable DataTables

### LINKED ISSUE #56079: DataTable should incorporate Theming
Following up on these two To-Dos:

https://github.com/flutter/flutter/blob/368da5bb5962db7be40439f416bd12176e2b3abf/packages/flutter/lib/src/material/data_table.dart#L546

https://github.com/flutter/flutter/blob/368da5bb5962db7be40439f416bd12176e2b3abf/packages/flutter/lib/src/material/data_table.dart#L599

By default the table's colors and text styles should be configured in terms of the Theme's colorScheme and textTheme. There shouldn't be brightness-specific code here or hardwired colors, etc. Additionally, there should be a DataTableTheme, so that one can configure the appearance of tables in general.

### PR DESCRIPTION
**Pull request by perclasson**
## Description

This PR adds a new DataTableTheme class to allow for themable DataTables across an application.

It also aligns the data table with Material specs by:
- To allow for a `DataTable.decoration` with border around the table, without having a double border at the bottom, we do not set the border at the bottom of the `DataTable` by default (use `DataTable.showBottomDivider` to still have it).
- Change heading text to be `subtitle2` by default.
- Change cell text to be `bodyText2` by default.
- Let sort icon color be defined by any theme set instead of hard coding it to a black and white color.

## Screenshots
| PaginatedDataTable before | PaginatedDataTable after |
| :---: | :---: |
|  |  |
|  |  |

## Related Issues

Closes https://github.com/flutter/flutter/issues/56079

Follow up issues:

- Make sure the selected color follows Material spec guidelines, and that tristate icon is correct: https://github.com/flutter/flutter/issues/64314
- Add theming class for `PaginatedDataTable` (will create new issue when more information has been gathered on what is needed)

## Tests

I added the following tests:

- DataTable set show bottom divider
- DataTableThemeData copyWith, ==, hashCode basics
- DataTableThemeData defaults
- Default DataTableThemeData debugFillProperties
- DataTableThemeData implements debugFillProperties
- DataTable is themeable
- DataTable properties are taken over the theme values

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by JoseAlba — COMMENTED** [review]
In data_table.dart you are making a variable effectiveHorizontalMargin and all the other variables are effective... I think it would make more sense to replace effectiveHorizontalMargin with _defaultHorizontalMargin. Because this is a private variable and it makes it easier to read when someone is going through the code.

I think it is worth getting a second opinion on this though.



 final double effectiveHorizontalMargin = horizontalMargin
      ?? themeData.dataTableTheme?.horizontalMargin
      ?? DataTable.defaultHorizontalMargin;

**Review by JoseAlba — APPROVED** [review]
```
██╗      ██████╗ ████████╗███╗   ███╗
██║     ██╔════╝ ╚══██╔══╝████╗ ████║
██║     ██║  ███╗   ██║   ██╔████╔██║
██║     ██║   ██║   ██║   ██║╚██╔╝██║
███████╗╚██████╔╝   ██║   ██║ ╚═╝ ██║
╚══════╝ ╚═════╝    ╚═╝   ╚═╝     ╚═╝
```

**Review by rami-a — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/data_table.dart [hunk 1]
```diff
@@ -475,31 +478,110 @@ class DataTable extends StatelessWidget {
   /// row is selectable.
   final ValueSetter<bool> onSelectAll;
 
+  /// The background and border decoration for the table.
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:481]: I looked at BoxDecoration and the Ink widget, neither of them has any templates, and they are not more specific about documentation the decoration parameter, so I'll leave this out.

File: packages/flutter/lib/src/material/data_table.dart [hunk 2]
```diff
@@ -475,31 +478,110 @@ class DataTable extends StatelessWidget {
...
+
+  /// The color for the data rows.
+  ///
+  /// By default, the color is transparent unless selected. Selected rows has
+  /// a grey translucent color. To set a different color for individual rows,
+  /// see [DataRow.color].
+  ///
+  /// The effective color can depend on the [MaterialState] state, if the
+  /// row is selected, pressed, hovered, focused, disabled or enabled. The
+  /// color is painted as an overlay to the row. To make sure that the row's
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:494]: That is correct

File: packages/flutter/lib/src/material/data_table.dart [hunk 3]
```diff
@@ -475,31 +478,110 @@ class DataTable extends StatelessWidget {
...
+  final MaterialStateProperty<Color> dataRowColor;
+
   /// The height of each row (excluding the row that contains column headings).
   ///
-  /// This value defaults to kMinInteractiveDimension to adhere to the Material
+  /// This value defaults to [kMinInteractiveDimension] to adhere to the Material
   /// Design specifications.
   final double dataRowHeight;
 
+  /// The text style for data rows.
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:521]: I could not find a useful macro or template for TextStyle. I think by going to TextStyle directly they can find useful enough information.

File: packages/flutter/lib/src/material/data_table.dart [hunk 4]
```diff
@@ -475,31 +478,110 @@ class DataTable extends StatelessWidget {
...
+  ///
+  /// By default, the text style is [TextTheme.bodyText2].
+  final TextStyle dataTextStyle;
+
+  /// The color for the heading row.
+  ///
+  /// By default, the color is transparent.
+  ///
+  /// The effective color can depend on the [MaterialState] state, if the
+  /// row is pressed, hovered, focused. The color is painted as an overlay
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:531]: Yes if its sortable.

File: packages/flutter/lib/src/material/data_table.dart [hunk 5]
```diff
@@ -475,31 +478,110 @@ class DataTable extends StatelessWidget {
...
   ///
-  /// This value defaults to 56.0 to adhere to the Material Design specifications.
+  /// This value defaults to [DataTable.defaultColumnSpacing] to adhere to the
+  /// Material Design specifications.
   final double columnSpacing;
 
+  /// The padding between header text and the sort icon.
+  ///
+  /// By default the padding is [DataTable.defaultSortArrowPadding].
+  final EdgeInsetsGeometry sortIconPadding;
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:583]: I will remove this one parameter, I thought initially I would need it for internal usages but probably not needed.

File: packages/flutter/lib/src/material/data_table.dart [hunk 6]
```diff
@@ -551,29 +645,42 @@ class DataTable extends StatelessWidget {
...
+    BuildContext context,
     Color activeColor,
     bool checked,
     VoidCallback onRowTap,
     ValueChanged<bool> onCheckboxChanged,
     MaterialStateProperty<Color> overlayColor,
   }) {
+    final ThemeData themeData = Theme.of(context);
+    final double effectiveHorizontalMargin = horizontalMargin
+      ?? themeData.dataTableTheme?.horizontalMargin
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:675]: Good catch

File: packages/flutter/test/material/data_table_test.dart [hunk 7]
```diff
@@ -571,7 +576,7 @@ void main() {
 
     // default checkbox padding
     checkbox = find.byType(Checkbox).first;
-    padding = find.ancestor(of: checkbox, matching: find.byType(Padding));
+    padding = find.ancestor(of: checkbox, matching: find.byType(Padding)).first;
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/test/material/data_table_test.dart:579]: It is due to the change to how the padding is added to the sort icon. Now that I remove that option, I can revert this change actually.

File: packages/flutter/test/material/data_table_test.dart [hunk 8]
```diff
@@ -1195,6 +1249,7 @@ void main() {
     const Color pressedColor = Color(0xff4caf50);
     Widget buildTable() {
       return DataTable(
+        dataTextStyle: const TextStyle(fontSize: 13.0),
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/test/material/data_table_test.dart:1252]: I changed the text style for the data text, so to keep the test the same I could either update the variables that we `expect` or set the style to be the same as before.

Seems like it's confusing to just add the textStyle so I'll go ahead and update the values instead.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 9]
```diff
@@ -1673,6 +1690,7 @@ class ThemeData with Diagnosticable {
     properties.add(DiagnosticsProperty<TextButtonThemeData>('textButtonTheme', textButtonTheme, defaultValue: defaultData.textButtonTheme, level: DiagnosticLevel.debug));
     properties.add(DiagnosticsProperty<ElevatedButtonThemeData>('elevatedButtonTheme', elevatedButtonTheme, defaultValue: defaultData.elevatedButtonTheme, level: DiagnosticLevel.debug));
     properties.add(DiagnosticsProperty<OutlinedButtonThemeData>('outlinedButtonTheme', outlinedButtonTheme, defaultValue: defaultData.outlinedButtonTheme, level: DiagnosticLevel.debug));
+    properties.add(DiagnosticsProperty<DataTableThemeData>('dataTableTheme', dataTableTheme, defaultValue: defaultData.dataTableTheme, level: DiagnosticLevel.debug));
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/theme_data.dart:1693]: I think that is out of the scope of this pr

File: packages/flutter/lib/src/material/data_table.dart [hunk 10]
```diff
@@ -517,6 +599,18 @@ class DataTable extends StatelessWidget {
...
+  ///
+  /// Must be non-null and greater than or equal to zero.
+  /// This value defaults to 1.0.
+  final double dividerThickness;
+
+  /// Whether a divider at the bottom of the table is displayed.
+  ///
+  /// By default, a divider not is shown at the bottom to allow for a border
+  /// around the table set with [DataTable.decoration].
+  final bool showBottomDivider;
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:612]: Done!

File: packages/flutter/lib/src/material/data_table.dart [hunk 11]
```diff
@@ -1035,10 +1162,9 @@ class _SortArrowState extends State<_SortArrow> with TickerProviderStateMixin {
         transform: Matrix4.rotationZ(_orientationOffset + _orientationAnimation.value)
                              ..setTranslationRaw(0.0, _arrowIconBaselineOffset, 0.0),
         alignment: Alignment.center,
-        child: Icon(
+        child: const Icon(
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:1165]: Yes, let me add a screenshot to the description

File: packages/flutter/lib/src/material/data_table_theme.dart [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    this.dataRowHeight,
+    this.dataTextStyle,
+    this.headingRowColor,
+    this.headingRowHeight,
+    this.headingTextStyle,
+    this.horizontalMargin,
+    this.columnSpacing,
+  });
+
+  /// {@macro flutter.material.dataTable.decoration}
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/data_table_theme.dart:48]: Looks good! Only issue is I think the macros include the "Defaults by" but the DataTableThemeDate properties should be null by default

**Review by guidezpl — APPROVED** [review]
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

File: packages/flutter/lib/src/material/data_table_theme.dart (hunk 12 above)
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table_theme.dart:48]: True I will change the templates a bit to exclude this. I think for `headingRowColor` and `dataRowColor` I can exclude this part. For headingRowColor it is null and not used by default anyways and for dataRowColor the default should be updated in https://github.com/flutter/flutter/issues/64314 to be the primary color.

File: packages/flutter/lib/src/material/data_table.dart [hunk 13]
```diff
@@ -414,21 +414,23 @@ class DataTable extends StatelessWidget {
...
+    this.dataRowHeight,
+    this.dataTextStyle,
+    this.headingRowColor,
+    this.headingRowHeight,
+    this.headingTextStyle,
+    this.horizontalMargin,
+    this.columnSpacing,
     this.showCheckboxColumn = true,
+    this.showBottomBorder = false,
     this.dividerThickness = 1.0,
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/data_table.dart:428]: Why isn't this part of the theme?

**Review by HansMuller — COMMENTED** [review]
This looks good, just some small stuff.

File: packages/flutter/test/material/data_table_theme_test.dart [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+    expect(theme.data.horizontalMargin, null);
+    expect(theme.data.columnSpacing, null);
+  });
+
+  testWidgets('Default DataTableThemeData debugFillProperties', (WidgetTester tester) async {
+    final DiagnosticPropertiesBuilder builder = DiagnosticPropertiesBuilder();
+    const DataTableThemeData().debugFillProperties(builder);
+
+    final List<String> description = builder.properties
+        .where((DiagnosticsNode node) => !node.isFiltered(DiagnosticLevel.info))
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/test/material/data_table_theme_test.dart:46]: Done!

File: packages/flutter/lib/src/material/data_table_theme.dart [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+      dataTextStyle: dataTextStyle ?? this.dataTextStyle,
+      headingRowColor: headingRowColor ?? this.headingRowColor,
+      headingRowHeight: headingRowHeight ?? this.headingRowHeight,
+      headingTextStyle: headingTextStyle ?? this.headingTextStyle,
+      horizontalMargin: horizontalMargin ?? this.horizontalMargin,
+      columnSpacing: columnSpacing ?? this.columnSpacing,
+    );
+  }
+
+  /// Linearly interpolate between two [DataTableThemeData].
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table_theme.dart:102]: Done!

File: packages/flutter/lib/src/material/data_table.dart [hunk 16]
```diff
@@ -475,27 +477,117 @@ class DataTable extends StatelessWidget {
   /// row is selectable.
   final ValueSetter<bool> onSelectAll;
 
+  /// {@template flutter.material.dataTable.decoration}
+  /// The background and border decoration for the table.
+  /// {@endtemplate}
+  ///
+  /// By default there is no decoration.
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:484]: I think in this case we will not need to have different box decorations depending on the state. The focus or selected state will only be handled by the different rows and the whole DataTable never really has any state.

File: packages/flutter/lib/src/material/data_table.dart (hunk 13 above)
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:428]: Good point, I added dividerThickness to the theme now.

File: packages/flutter/lib/src/material/data_table_theme.dart [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+/// Descendant widgets obtain the current [DataTableThemeData] object
+/// using `DataTableTheme.of(context)`. Instances of
+/// [DataTableThemeData] can be customized with
+/// [DataTableThemeData.copyWith].
+///
+/// Typically a [DataTableThemeData] is specified as part of the
+/// overall [Theme] with [ThemeData.dataTableTheme].
+///
+/// All [DataTableThemeData] properties are `null` by default. When
+/// null, the [DataTable]'s build method provides defaults.
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table_theme.dart:27]: Done!

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by tsinis** [comment]
Just my two cents: DataTable does not respect font colors in CupertinoApp, when dark theme is used. Here is a simple code: 
https://dartpad.dev/501d270700d5544d2292d509b86f79af
DataTable should be Material independent IMHO.

**Comment by perclasson** [comment]
> Just my two cents: DataTable does not respect font colors in CupertinoApp, when dark theme is used. Here is a simple code:
> https://dartpad.dev/501d270700d5544d2292d509b86f79af
> DataTable should be Material independent IMHO.

Please feel free to create an issue with the bug you're seeing, as it's not related to this PR. That way we can address and prioritize the issue and make sure it gets looked at.

**Pull request merged**